pyxcp/transport/eth.py

```diff
@@ -69,37 +75,62 @@ def listen(self):
 
         while True:
             try:
-                if self.closeEvent.isSet() or self.sock.fileno() == -1:
+                if close_event_set() or socket_fileno() == -1:
                     return
-                sel = self.selector.select(0.1)
+                sel = select(0.1)
                 for _, events in sel:
-                    if events & selectors.EVENT_READ:
+                    if events & EVENT_READ:
                         if use_tcp:
-                            header = bytearray()
 
-                            while len(header) < HEADER_SIZE:
-                                new_bytes = sock_recv(
-                                    HEADER_SIZE - len(header))
-                                header.extend(new_bytes)
+                            # first try to get the header in one go
+                            # if we are lucky this will avoid creating a bytearray
+                            # and extending it
+                            header = sock_recv(HEADER_SIZE)
+                            size = len(header)
+                            if size != HEADER_SIZE:
```

Comment:
I'm not a big fan of replacing the condition of a while loop with an infinite loop, an if statement and a break. Are you sure that's an optimization? Either way, it's a comparison + a conditional branch - no?

I think the essence of your optimization is to prevent the creation of the bytearray. Isn't it enough to replace `header = bytearray()` with `header = sock_recv(HEADER_SIZE)` and then wrapping the while loop with

```
if len(header) != HEADER_SIZE:
    header = bytearray(header)
```

Same thing for the response section.

Reply:
That's fair, the while loop looks more idiomatic. I'll have a go a reverting it tomorrow